core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/HealthMonitor.scala

```diff
@@ -0,0 +1,238 @@
+import scala.util.control.NonFatal
+
+/**
+  * Created by rtitle on 5/17/17.
+  */
+object HealthMonitor {
+  val DefaultFutureTimeout = 1 minute
+  val DefaultStaleThreshold = 15 minutes
+
+  val OkStatus = SubsystemStatus(true, None)
+  val UnknownStatus = SubsystemStatus(false, Some(List("Unknown status")))
+  def failedStatus(message: String) = SubsystemStatus(false, Some(List(message)))
+
+  // Actor API:
+
+  /** Triggers subsystem checking */
+  case object CheckAll
+  /** Stores status for a particular subsystem */
+  case class Store(subsystem: Subsystem, status: SubsystemStatus)
+  /** Retrieves current status and sends back to caller */
+  case object GetCurrentStatus
+
+  def props(slickDataSource: SlickDataSource, googleServicesDAO: GoogleServicesDAO, googlePubSubDAO: GooglePubSubDAO, userDirectoryDAO: UserDirectoryDAO, methodRepoDAO: MethodRepoDAO,
+            groupsToCheck: Seq[String], topicsToCheck: Seq[String], bucketsToCheck: Seq[String],
+            futureTimeout: FiniteDuration = DefaultFutureTimeout, staleThreshold: FiniteDuration = DefaultStaleThreshold): Props =
+    Props(new HealthMonitor(slickDataSource, googleServicesDAO, googlePubSubDAO, userDirectoryDAO, methodRepoDAO, groupsToCheck, topicsToCheck, bucketsToCheck, futureTimeout, staleThreshold))
+}
+
+class HealthMonitor private (val slickDataSource: SlickDataSource, val googleServicesDAO: GoogleServicesDAO, val googlePubSubDAO: GooglePubSubDAO, val userDirectoryDAO: UserDirectoryDAO, val methodRepoDAO: MethodRepoDAO,
+                             val groupsToCheck: Seq[String], val topicsToCheck: Seq[String], val bucketsToCheck: Seq[String],
+                             val futureTimeout: FiniteDuration, val staleThreshold: FiniteDuration) extends Actor with LazyLogging {
+  // Use the execution context for this actor's dispatcher for all asynchronous operations.
+  // We define a separate execution context (a fixed thread pool) for health checking to
+  // not interfere with user facing operations.
+  import context.dispatcher
+
+  /**
+    * Contains each subsystem status along with a timestamp of when the entry was made.
+    * Initialized with unknown status.
+    */
+  var data: Map[Subsystem, (SubsystemStatus, Long)] = {
+    val now = System.currentTimeMillis
+    AllSubsystems.map(_ -> (UnknownStatus, now)).toMap
+  }
+
+  override def receive: Receive = {
+    case CheckAll => checkAll
+    case Store(subsystem, status) => store(subsystem, status)
+    case GetCurrentStatus => sender ! getCurrentStatus
+  }
+
+  private def checkAll: Unit = {
+    List(
+      (Agora, checkAgora),
+      (Cromwell, checkCromwell),
+      (Database, checkDB),
+      (GoogleBilling, checkGoogleBilling),
+      (GoogleBuckets, checkGoogleBuckets),
+      (GoogleGenomics, checkGoogleGenomics),
+      (GoogleGroups, checkGoogleGroups),
+      (GooglePubSub, checkGooglePubsub),
+      (LDAP, checkLDAP)
+    ).foreach(processSubsystemResult _ tupled)
```

Comment:
Actually `tupled` is a method on `scala.Function2` which returns a function which takes the parameters as a tuple (instead of individual args).
See: http://www.scala-lang.org/api/2.9.2/scala/Function2.html

For example:
```
scala> def foo(a: Int, b: Int) = a + b
foo: (a: Int, b: Int)Int

scala> :t foo _
(Int, Int) => Int

scala> :t (foo _).tupled
((Int, Int)) => Int
```

So with parens this would be written as `.foreach((processSubsystemResult _).tupled)`. Maybe that is more clear...